processing/src/main/java/org/apache/druid/query/aggregation/AggregatorFactory.java

```diff
@@ -214,19 +216,42 @@ public AggregatorFactory getMergingFactory(AggregatorFactory other) throws Aggre
    * {@link #deserialize} and the type accepted by {@link #combine}. However, it is *not* necessarily the same type
    * returned by {@link #finalizeComputation}.
    *
-   * If the type is complex (i.e. not a simple, numeric {@link org.apache.druid.segment.column.ValueType}) then there
+   * If the type is complex (i.e. not a simple, numeric {@link ValueType}) then there
    * must be a corresponding {@link org.apache.druid.segment.serde.ComplexMetricSerde} which was registered with
    * {@link org.apache.druid.segment.serde.ComplexMetrics#registerSerde} using this type name.
    *
-   * If you need a ValueType enum corresponding to this aggregator, a good way to do that is:
-   *
-   * <pre>
-   *   Optional.ofNullable(GuavaUtils.getEnumIfPresent(ValueType.class, aggregator.getTypeName()))
-   *           .orElse(ValueType.COMPLEX);
-   * </pre>
+   * If you need a ValueType enum corresponding to this aggregator, use {@link #getTypeName} instead.
    */
   public abstract String getTypeName();
 
+  /**
+   * Get the type name for the 'finalized' type for this aggregator, i.e. the type of the value returned by
+   * {@link #finalizeComputation}. This may be the same as or different than the types expected in {@link #deserialize}
+   * and {@link #combine}.
+   *
+   * If you need a ValueType enum corresponding to this aggregator, use {@link #getFinalizedType} instead.
+   */
+  public String getFinalizedTypeName()
+  {
+    return getTypeName();
```

Comment:
can we add a blurb recommending this method be explicitly overridden by complex AggregatorFactory impls as the default here is likely wrong for those.

Reply:
This might actually be better as an abstract method so that all `AggregatorFactory` must explicitly specify the finalized type, but I was somewhat worried about it being a bit disruptive. I think it is worth discussing if abstract would be better, but yeah at least will add javadocs.

Reply:
actually, this one can also be removed in we just need to have `getFinalizedType()` and it having default impl `return getType()`  .
I didn't suggest making it abstract because it does work correctly for all of aggregators dealing with primitives, so default is ok specially because "wrong" behavior doesn't appear to cause any correctness issue for query processing.

Reply:
I think maybe if we remove both `getFinalizedTypeName` from the `AggregatorFactory` and `getTypeName` from the `PostAggregator`, we should also maybe consider renaming `AggregatorFactory.getTypeName` to be `AggregatorFactory.getComplexTypeName` and ensure it is only still called for getting a `ComplexMetricSerde` for the aggregator. I'll look into this when I get back to this branch, but a rename like that might be sort of disruptive to extension writers.

Reply:
>  ensure it is only still called for getting a ComplexMetricSerde for the aggregator.

that sounds ok, with `ValueType getType()` in there, `AggregatorFactory.getTypeName()` is only used to find  right `ComplexMetricSerde` object .. you are right  that changing the name might be disruptive, so adding the  right requirement in javadoc would be a good compromise.... or maybe you would find a better alternative.

Reply:
I've transitioned `AggregatorFactory.getTypeName` to be used exclusively for complex type serde lookup, and remove all of the other type name functions. All `AggregatorFactory` implementations now must explicitly implement `getType` instead of a default that tries to convert it into a `ValueType` enum from the result of `getTypeName`, and I removed `getTypeName` from all non-complex agg factories in favor of a default implementation that throws an exception (to ensure this method isn't called inappropriately).

Since `getType` is now abstract, this PR is sort of disruptive to agg extensions, so it might be also ok to change `getTypeName` to `getComplexTypeName` since a code change will be required anyway. 

Alternatively, I could restore the default implementation of `getType` that uses the output of `getTypeName` to translate into a `ValueType` if this is too much. I removed it as much so that I would go through and provide explicit implementations for all existing aggregators using errors to hunt them all down.

Reply:
Fwiw, I'm in favor of requiring aggregators to explicitly implement these functions — especially if 90% of people would be fine with the default! That means it would be really easy for the 10% that need to override it to forget to do so. In general it's good to make it easy to do things right, not easy to forget important stuff.

Reply:
I agree, `getType` and `getFinalizedType` are now both `abstract`, and I also have renamed `getTypeName` to be `getComplexTypeName` since it didn't seem much more disruptive to rename an existing method on top of having to implement the 2 new methods.